test/shared/RoutingTestHelpers.sol

```diff
@@ -0,0 +1,115 @@
+        positionManager = new PoolModifyLiquidityTest(manager);
+
+        MockERC20[] memory tokens = deployTokensMintAndApprove(4);
+
+        currency0 = Currency.wrap(address(tokens[0]));
+        currency1 = Currency.wrap(address(tokens[1]));
+        currency2 = Currency.wrap(address(tokens[2]));
+        currency3 = Currency.wrap(address(tokens[3]));
+
+        key0 = createPoolWithLiquidity(currency0, currency1, address(0));
+        key1 = createPoolWithLiquidity(currency1, currency2, address(0));
+        key2 = createPoolWithLiquidity(currency2, currency3, address(0));
+    }
+
+    function deployTokensMintAndApprove(uint8 count) internal returns (MockERC20[] memory) {
+        MockERC20[] memory tokens = deployTokens(count, 2 ** 128);
+        for (uint256 i = 0; i < count; i++) {
+            tokens[i].approve(address(router), type(uint256).max);
+        }
+        return tokens;
+    }
+
+    function createPoolWithLiquidity(Currency currencyA, Currency currencyB, address hookAddr)
+        internal
+        returns (PoolKey memory _key)
+    {
+        if (Currency.unwrap(currencyA) > Currency.unwrap(currencyB)) (currencyA, currencyB) = (currencyB, currencyA);
+        _key = PoolKey(currencyA, currencyB, 3000, 60, IHooks(hookAddr));
+
+        manager.initialize(_key, SQRT_PRICE_1_1, ZERO_BYTES);
+        MockERC20(Currency.unwrap(currencyA)).approve(address(positionManager), type(uint256).max);
+        MockERC20(Currency.unwrap(currencyB)).approve(address(positionManager), type(uint256).max);
+        positionManager.modifyLiquidity(_key, IPoolManager.ModifyLiquidityParams(-887220, 887220, 200 ether, 0), "0x");
+    }
+
+    function _getExactInputParams(Currency[] memory _tokenPath, uint256 amountIn)
+        internal
+        pure
+        returns (IV4Router.ExactInputParams memory params)
+    {
+        PathKey[] memory path = new PathKey[](_tokenPath.length - 1);
+        for (uint256 i = 0; i < _tokenPath.length - 1; i++) {
+            path[i] = PathKey(_tokenPath[i + 1], 3000, 60, IHooks(address(0)), bytes(""));
+        }
+
+        params.currencyIn = _tokenPath[0];
+        params.path = path;
+        params.amountIn = uint128(amountIn);
+        params.amountOutMinimum = 0;
+    }
+
+    function _getExactOutputParams(Currency[] memory _tokenPath, uint256 amountOut)
+        internal
+        pure
+        returns (IV4Router.ExactOutputParams memory params)
+    {
+        PathKey[] memory path = new PathKey[](_tokenPath.length - 1);
+        for (uint256 i = _tokenPath.length - 1; i > 0; i--) {
+            path[i - 1] = PathKey(_tokenPath[i - 1], 3000, 60, IHooks(address(0)), bytes(""));
+        }
+
+        params.currencyOut = _tokenPath[_tokenPath.length - 1];
+        params.path = path;
+        params.amountOut = uint128(amountOut);
+        params.amountInMaximum = type(uint128).max;
+    }
+
+    function _finalizePlan(Currency inputCurrency, Currency outputCurrency, address recipient) internal {
```

Comment:
wanna add this to Planner.sol? maybe just overload planner.finalize? or you could call it like planner.finalizeSwap()?